app/src/main/java/sw19/moring03/paint/tools/FillTool.java

```diff
@@ -0,0 +1,103 @@
+package sw19.moring03.paint.tools;
+
+import android.graphics.Canvas;
+import android.graphics.Color;
+import android.graphics.Paint;
+import sw19.moring03.paint.utils.PointF;
+
+import java.util.ArrayList;
+import java.util.Stack;
+
+public class FillTool extends Tools {
+
+    private int[] frameBuffer;
+    private int bufferWidth;
+    private int bufferHeight;
+
+    public FillTool() {
+        points = new ArrayList<>();
+        strokeWidth = 5;
+        color = Color.BLACK;
+    }
+
+    public FillTool(int col, int[] frameBuffer, int bufferWidth, int bufferHeight) {
+        points = new ArrayList<>();
+        this.frameBuffer = frameBuffer;
+        this.bufferWidth = bufferWidth;
+        this.bufferHeight = bufferHeight;
+        color = col;
+    }
+
+    @Override
+    public boolean draw(Canvas canvas, Paint paint) {
+        if (points.size() == 1 && frameBuffer != null) {
+            PointF start = points.remove(0);
+            int scanColor = frameBuffer[(int)start.y * bufferWidth + (int)start.x];
+            scanlineFloodFillStack((int)start.x, (int)start.y, scanColor);
+            this.frameBuffer = null;
+        }
+
+        if (points.size() % 2 != 0 || points.isEmpty()) {
```

Comment:
change the two statements so yo check for empty first. Just a bit prettier and more efficient imo. But not necessary to change

Reply:
It also makes the code more correct, since `0 % 2` also evaluates to `true`. So we jump over this condition even if there are no points in the list.

Fixed with commit  7225edaa72.